tools/lsp/properties.rs

```diff
@@ -0,0 +1,141 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+use i_slint_compiler::diagnostics::Spanned;
+use i_slint_compiler::object_tree::{Element, ElementRc};
+
+#[derive(serde::Deserialize, serde::Serialize)]
+pub(crate) struct PropertyInformation {
+    name: String,
+    type_name: String,
+    declared_at: Option<(String, u32)>,
+    defined_at: Option<(u32, u32)>, // Range in the elements source file!
+}
+
+#[derive(serde::Deserialize, serde::Serialize)]
+pub(crate) struct QueryPropertyResponse {
+    properties: Vec<PropertyInformation>,
+    source_file: Option<String>,
+}
+
+// This gets defined accessibility properties...
+fn get_reserved_properties(result: &mut Vec<PropertyInformation>) {
+    result.extend(i_slint_compiler::typeregister::reserved_properties().map(|p| {
+        PropertyInformation {
+            name: p.0.to_string(),
+            type_name: format!("{}", p.1),
+            declared_at: None,
+            defined_at: None,
+        }
+    }))
+}
+
+fn source_file(element: &Element) -> Option<String> {
+    element.source_file().map(|sf| sf.path().to_string_lossy().to_string())
+}
+
+fn get_element_properties(element: &Element, result: &mut Vec<PropertyInformation>) {
+    let file = source_file(element);
+
+    for (name, value) in &element.property_declarations {
+        let declared_at = file.as_ref().and_then(|file| {
+            value.type_node().map(|n| n.text_range().start().into()).map(|p| (file.clone(), p))
+        });
+        result.push(PropertyInformation {
+            name: name.clone(),
+            type_name: format!("{}", value.property_type),
+            declared_at,
+            defined_at: None,
+        });
+    }
+}
+
+fn insert_property_definition_range(
+    property: &str,
+    properties: &mut Vec<PropertyInformation>,
+    range: (u32, u32),
+) {
+    let index = properties
+        .binary_search_by(|p| (&p.name[..]).cmp(&property))
+        .expect("property must be known");
+    properties[index].defined_at = Some(range);
+}
+
+fn insert_property_definitions(element: &Element, properties: &mut Vec<PropertyInformation>) {
+    let node = if let Some(node) = element.node.as_ref() {
+        node
+    } else {
+        return;
+    };
+
+    for (k, v) in &element.bindings {
+        if let Some(span) = &v.borrow().span {
+            let offset: u32 = span.span().offset.try_into().unwrap_or(u32::MAX);
+            if element.source_file().map(|sf| sf.path())
+                != span.source_file.as_ref().map(|sf| sf.path())
```

Comment:
Since this happens from time to time, I wonder if `SourceFile` should be a new type instead of an alias, so that `PartialEq` can be implemented for it?

Reply:
Out of scope for this PR.